Fix duplicate metrics in AWX subsystem_metrics

cbcoutinho opened the pull request:
##### SUMMARY
Fixes #15179.

The original PR (#14775) that introduced this bug included a non-empty `_METRICSLIST` in the `Metrics` class which was subsequently subclassed twice - by the `DispatcherMetrics` and `CallbackMetrics`, causing duplicate metrics. This issue doesn't show up immediately upon startup. Once a workflow starts, this metrics appears twice in the `/metrics` endpoint.

This PR renames the `subsystem_*` metrics into separate metrics for both the `CallbackReceiver` and the `Dispatcher`.

This is technically a breaking change in the metrics API because three metrics have been removed, being replaced by three metrics for `CallbackReceiver` and three for the `Dispatcher`

##### ISSUE TYPE
 - Bug fix in results from `/metrics` endpoint

Prometheus logs:
Without this change, Prometheus drops the duplicate metrics and logs errors:

```
prometheus-observability-kube-prometh-prometheus-0 prometheus time=2025-04-27T11:25:48.202Z level=WARN source=scrape.go:1884 msg="Error on ingesting samples with different value but same timestamp" component="scrape manager" scrape_pool=serviceMonitor/awx/awx-web-monitor/0 target=http://10.1.76.143:8052/api/v2/metrics/ num_dropped=3
```

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Other

##### AWX VERSION

```
awx: 24.6.2.dev314+g52da7b86df
```


##### ADDITIONAL INFORMATION
See https://github.com/ansible/awx/issues/15179#issuecomment-2817296892 for context

cbcoutinho committed:
Fix duplicate metrics in AWX subsystem_metrics

Separate out operation subsystem metrics to fix duplicate error

Remove unnecessary comments

Revert to single subsystem_metrics_* metric with labels

Format via black

AlanCoding commented:
I checked the logs from one of the awx_collection tests and it gave this:

```
2025-05-16 20:42:23,602 ERROR    [-] awx.main.commands.run_callback_receiver Exception on worker 3, reconnecting: 
Traceback (most recent call last):
  File "/awx_devel/awx/main/dispatch/worker/base.py", line 315, in work_loop
    body = self.read(queue)
           ^^^^^^^^^^^^^^^^
  File "/awx_devel/awx/main/dispatch/worker/callback.py", line 113, in read
    self.record_read_metrics()
  File "/awx_devel/awx/main/dispatch/worker/callback.py", line 123, in record_read_metrics
    self.subsystem_metrics.pipe_execute()
  File "/awx_devel/awx/main/analytics/subsystem_metrics.py", line 279, in pipe_execute
    self.METRICS['subsystem_metrics_pipe_execute_seconds'].inc(duration_to_save)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'subsystem_metrics_pipe_execute_seconds'
```

I think those artifacts are available publicly? You have to click on the check output, and then Summary, and then down at the bottom it has artifacts. That lets you get the logs which often tell what actually went wrong, like the above trace

cbcoutinho commented:
Hi @AlanCoding thanks - I've updated my approach based on the failing tests. My initial naive solution was to remove the metrics from one of the subsystems to eliminate the duplicates, but that isn't sufficient.

In my recent changes, I renamed the `subsystem_*` metrics into separate metrics for both the _Callback Receiver_ and the _Dispatcher_.

This is technically a breaking change in the metrics API because three metrics have been removed, being replaced by three metrics for _Callback Receiver_ and three for the _Dispatcher_.

Can you please enable the checks again?

Sispheor commented:
Hello there !
We would love to see that fixed merged 😁. What is the status?

cbcoutinho commented:
Hi @AlanCoding can you please re-run the tests?

cbcoutinho commented:
Friendly ping @AlanCoding @chrismeyersfsu @kdelee

thedoubl3j commented:
branch is out of date so I should not have kicked CI, rebasing the branch and that should kick off CI again.

pb82 commented:
@cbcoutinho instead of renaming the metrics, could we just add a label with the name of the subsystem? That would also resolve the duplication and not cause an API change.

cbcoutinho commented:
Hi @pb82, great idea - I've refactored the PR such that metrics optionally can add a sub-system to the metric and avoids breaking the metrics API

pb82 commented:
@cbcoutinho I've verified your change, here are the before and after results from Prometheus:

<img width="1908" height="302" alt="metrics_issue_before" src="https://github.com/user-attachments/assets/5b49f9c9-e459-4d06-baeb-afe1b51a39a4" />
<img width="1910" height="401" alt="metrics_issue_after" src="https://github.com/user-attachments/assets/b8735b13-f8a8-4f83-bd49-9ad2cf2b04c1" />

Does that look right to you? Looks good to me, all metrics are there and can be distinguished with labels.

traytorous reviewed (approved):
LGTM!

cbcoutinho commented:
Thanks for confirming @pb82 , yes that's what I anticipate as well

pb82 commented:
@cbcoutinho only some linter issues left, you can probably resolve them by running `make black`

cbcoutinho commented:
@pb82 @traytorous I've formatted the files using black so the linting should now  pass

pb82 commented:
@cbcoutinho merged, thanks again for your work on this!
